File: examples/canonical-auth/go/internal/routes/me.go
package routes

import (
	"net/http"
	"time"

	"canonical-auth/internal/auth"
	"canonical-auth/internal/domain"
	"canonical-auth/internal/registry"

	"github.com/labstack/echo/v4"
)

type MeRoutes struct {
	Store *domain.InMemoryStore
}

type MeResponse struct {
	Sub           string                   `json:"sub"`
	Email         string                   `json:"email"`
	DisplayName   string                   `json:"display_name"`
	TenantID      *string                  `json:"tenant_id"`
	TenantName    string                   `json:"tenant_name,omitempty"`
	TenantRole    string                   `json:"tenant_role"`
	Groups        []string                 `json:"groups"`
	Permissions   []string                 `json:"permissions"`
	ACLVer        int                      `json:"acl_ver"`
	AllowedRoutes []registry.RouteMetadata `json:"allowed_routes"`
	GuideSections []string                 `json:"guide_sections"`
	IssuedAt      time.Time                `json:"issued_at"`
	ExpiresAt     time.Time                `json:"expires_at"`
}

func (r *MeRoutes) Me(c echo.Context) error {
	authCtx, ok := c.Get("auth").(*auth.AuthContext)
	if !ok {
		return echo.NewHTTPError(http.StatusInternalServerError, "missing auth context")
	}

	user, ok := r.Store.GetUserByID(authCtx.Sub)
	if !ok {
		return echo.NewHTTPError(http.StatusUnauthorized, "user not found")
	}

	var tenantName string
	if authCtx.TenantID != nil {
		tenantName = r.Store.GetTenantName(*authCtx.TenantID)
	}

	isSuperAdmin := authCtx.TenantRole == "super_admin"
	allowedRoutes := registry.GetAllowedRoutes(authCtx.Permissions, isSuperAdmin)

	var guideSections []string
	if !isSuperAdmin {
		guideSections = []string{"User Management", "Billing"}
	} else {
		guideSections = []string{}
	}

	groups := authCtx.Groups
	if groups == nil {
		groups = []string{}
	}
	permissions := authCtx.Permissions
	if permissions == nil {
		permissions = []string{}
	}

	resp := MeResponse{
		Sub:           authCtx.Sub,
		Email:         authCtx.Email,
		DisplayName:   user.DisplayName,
		TenantID:      authCtx.TenantID,
		TenantName:    tenantName,
		TenantRole:    authCtx.TenantRole,
		Groups:        groups,
		Permissions:   permissions,
		ACLVer:        authCtx.ACLVer,
		AllowedRoutes: allowedRoutes,
		GuideSections: guideSections,
		IssuedAt:      authCtx.IssuedAt,
		ExpiresAt:     authCtx.ExpiresAt,
	}

	return c.JSON(http.StatusOK, resp)
}

File: examples/canonical-auth/rust/src/routes/me.rs
use std::sync::Arc;
use axum::{
    extract::State,
    http::StatusCode,
    response::IntoResponse,
    Json,
};
use serde_json::json;

use crate::auth::claims::AuthContext;
use crate::domain::store::InMemoryStore;
use crate::registry::routes::get_allowed_routes;

pub async fn me_handler(
    auth_ctx: AuthContext,
    State(store): State<Arc<InMemoryStore>>,
) -> impl IntoResponse {
    let user = match store.get_user_by_id(&auth_ctx.sub) {
        Some(u) => u,
        None => return StatusCode::UNAUTHORIZED.into_response(),
    };

    let tenant_name = auth_ctx.tenant_id.as_ref()
        .and_then(|id| store.get_tenant_name(id))
        .unwrap_or("");

    let is_super_admin = auth_ctx.tenant_role == "super_admin";
    let allowed_routes = get_allowed_routes(&auth_ctx.permissions, is_super_admin);

    let guide_sections = if is_super_admin {
        vec![]
    } else {
        vec!["User Management", "Billing"]
    };

    let body = json!({
        "sub": auth_ctx.sub,
        "email": auth_ctx.email,
        "display_name": user.display_name,
        "tenant_id": auth_ctx.tenant_id,
        "tenant_name": tenant_name,
        "tenant_role": auth_ctx.tenant_role,
        "groups": auth_ctx.groups,
        "permissions": auth_ctx.permissions,
        "acl_ver": auth_ctx.acl_ver,
        "allowed_routes": allowed_routes,
        "guide_sections": guide_sections,
        "issued_at": auth_ctx.issued_at,
        "expires_at": auth_ctx.expires_at,
    });

    (StatusCode::OK, Json(body)).into_response()
}
